### PR TITLE
WI #1133 Make the diagnostics defined on FunctionDeclaration more precise

### LINKED ISSUE #1133: Diagnostics set on FunctionDeclaration should be more precise on specify which Token, SymbolReference, ... is in error
**FOLLOWING OF #1091 for FunctionDeclaration**

When you set an diagnostics on a FunctionDeclaration, all tokens of the Node are considered as error.
Problem is, if only a few token cause the error, the user must read the error message to understand what is wrong.

This issue can be extended to all cases where a syntax error is reported instead of a semantic error, for example :
- Line 541 in CrossChecker.cs
- Line 791 in CrossChecker.cs

After a few tests, only 2 cases remains:

# Duplicate procedure declaration
Declare the 2 procedure with the same signatures.
The error must be reported on the line that contains tokens "declare procedure XXXX".

But currently, the error is reported on the first token of the procedure which might be the formalized comment.

Code to reproduce:
```cobol
      %<<<
           toto
      %>>>     
       declare Foo private  
               input var1 pic X.
       end-declare.
       
       declare Foo private  
               input var1 pic X.
       end-declare.
```
Expected:
Error about duplicate declaration of Foo must appears on lines `       declare Foo private`

# Invalid parameter syntax 
Use invalid syntax in a parameter:
```
      %<<<
           toto
      %>>>     
       declare Foo private  
               input var1 pic X occurs 5.
       end-declare.
```
An error is reported on the formalized comment instead of the line which contains `var1`.

### PR DESCRIPTION
**Pull request by LANOYDO**
Fixes #1133 
First commit:
Duplicate procedure declaration & Invalid parameter syntax

Second commit:
"This issue can be extended to all cases where a syntax error is reported instead of a semantic error:
Line 541 in CrossChecker.cs
Line 791 in CrossChecker.cs"

### CODE REVIEW

File: TypeCobol/Compiler/Diagnostics/TypeCobolChecker.cs [hunk 1]
```diff
@@ -728,9 +731,9 @@ private static void CheckParameter([NotNull] ParameterDescriptionEntry parameter
             {
                 foreach (var condition in parameter.DataConditions)
                 {
-                    if (condition.LevelNumber?.Value != 88)
+                    if (condition.LevelNumber?.Value != 88 && condition.Name != null)
```
Comment (fm-117) [inline TypeCobol/Compiler/Diagnostics/TypeCobolChecker.cs]: I think we should not change the condition in the test here but only the error message. So you can handle the case of null name directly in the error message like `"Condition parameter \"" + (condition.Name ?? "<missing name>") + "\" must be level 88."` or add a specific test for the null name before this one and create a new diagnostic just for the null name case.
To be discussed with @smedilol.

File: TypeCobol.LanguageServer.Test/LSRTests/MultiCommentSyntaxColoring/input/MultiCommentSyntaxColoring.tlsp [hunk 2]
```diff
@@ -28,7 +28,7 @@
...
...bl/DVZZJCM58896372926830602866.cee\",\"diagnostics\":[{\"range\":{\"start\":{\"line\":30,\"character\":6},\"end\":{\"line\":30,\"character\":6}},\"severity\":1,\"code\":\"44\",\"source\":\"Page\",\"message\":\"Invalid use of '%'. Should be used with \\\"%<< ... (linebreak) %>>\\\" for MultiLine Comment or \\\"%<<< ... (linebreak) %>>>\\\" for Formalized Comment\"},{\"range\":{\"start\":{\"line\":33,\"character\":6},\"end\":{\"line\":33,\"character\":6}},\"severity\":1,\"code\":\"44\",\"source\":\"Page\",\"message\":\"Invalid use of '%'. Should be used with \\\"%<< ... (linebreak) %>>\\\" for MultiLine Comment or \\\"%<<< ... (linebreak) %>>>\\\" for Formalized Comment\"},{\"range\":{\"start\":{\"line\":31,\"character\":55},\"end\":{\"line\":31,\"character\":60}},\"severity\":1,\"code\":\"30\",\"source\":\"Page\",\"message\":\"Semantic error: Symbol MyData is not referenced\"},{\"range\":{\"start\":{\"line\":31,\"character\":21},\"end\":{\"line\":31,\"character\":60}},\"severity\":1,\"code\":\"27\",\"source\":\"Find the syntax diagram describing the statement in error in the language reference\",\"message\":\"Syntax error : Function not found 'TCOZDATE.CurrentDate'  output(?)\"}]}}"
```
Comment (fm-117) [inline TypeCobol.LanguageServer.Test/LSRTests/MultiCommentSyntaxColoring/input/MultiCommentSyntaxColoring.tlsp]: The last part `Function not found 'TCOZDATE.CurrentDate'` looks like another semantic error that is currently reported as a syntax error. Maybe we should update it too in this issue.

File: TypeCobol.LanguageServer.Test/LSRTests/MultiLineCommentIncScan/input/MultiLineCommentIncScan.tlsp [hunk 3]
```diff
@@ -62,43 +62,7 @@
...
-    },
-    {
-      "category": 0,
-      "message": "{\"jsonrpc\":\"2.0\",\"method\":\"textDocument/didChange\",\"params\":{\"textDocument\":{\"version\":0,\"uri\":\"file:/C:/Users/MAYANJE/AppData/Local/Temp/1/tcbl/DVZZJCM52032788062027908163.cee\"},\"uri\":\"file:/C:/Users/MAYANJE/AppData/Local/Temp/1/tcbl/DVZZJCM52032788062027908163.cee\",\"contentChanges\":[{\"range\":{\"start\":{\"line\":29,\"character\":9},\"end\":{\"line\":29,\"character\":10}},\"rangeLength\":1,\"text\":\"\"}]}}"
-    },
-    {
-      "category": 1,
-      "message": "{\"jsonrpc\":\"2.0\",\"method\":\"textDocument/publishDiagnostics\",\"params\":{\"uri\":\"file:///C:/Users/MAYANJE/AppData/Local/Temp/1/tcbl/DVZZJCM52032788062027908163.cee\",\"diagnostics\":[]}}"
```
Comment (fm-117) [inline TypeCobol.LanguageServer.Test/LSRTests/MultiLineCommentIncScan/input/MultiLineCommentIncScan.tlsp]: What happened to these steps in the original LSR test ?

File: TypeCobol.LanguageServer.Test/LSRTests/ReplacingSyntaxOff/input/replacingSyntaxOff.tlsp [hunk 4]
```diff
@@ -15,11 +15,11 @@
...
+      "message": "{\"jsonrpc\":\"2.0\",\"method\":\"textDocument/publishDiagnostics\",\"params\":{\"uri\":\"file:///C:/Users/BALLMA/AppData/Local/Temp/1/tcbl/TCOZBMX02229180100554248398.cee\",\"diagnostics\":[{\"range\":{\"start\":{\"line\":11,\"character\":23},\"end\":{\"line\":11,\"character\":41}},\"severity\":1,\"code\":\"27\",\"source\":\"Find the syntax diagram describing the statement in error in the language reference\",\"message\":\"Syntax error : Ambiguous reference to symbol CONVEA-TESTVEA-TEST \\r\\nSymbols found: DVZS0OSM::CONVEA::CONVEA-TESTVEA-TEST | DVZS0OSM::CONVEAL::CONVEA-TESTVEA-TEST\"},{\"range\":{\"start\":{\"line\":12,\"character\":23},\"end\":{\"line\":12,\"character\":42}},\"severity\":1,\"code\":\"30\",\"source\":\"Page\",\"message\":\"Semantic error: Symbol CONVEAL-TESTVEA-TEST is not referenced\"}]}}"
```
Comment (fm-117) [inline TypeCobol.LanguageServer.Test/LSRTests/ReplacingSyntaxOff/input/replacingSyntaxOff.tlsp]: Same remark here : ambiguous reference to a symbol sounds more like a Semantic error than a syntax error.

File: TypeCobol/Compiler/Diagnostics/TypeCobolChecker.cs [hunk 5]
```diff
@@ -631,9 +631,12 @@ public static void OnNode(FunctionDeclaration functionDeclaration)
             var headerNameURI = new URI(header.Name);
             var functions = functionDeclaration.SymbolTable.GetFunction(headerNameURI, functionDeclaration.Profile);
             if (functions.Count > 1)
-                DiagnosticUtils.AddError(functionDeclaration,
+            {
+                Token declareToken = header.ConsumedTokens.First(t => t.TokenType == TokenType.DECLARE);
```
Comment (fm-117) [inline TypeCobol/Compiler/Diagnostics/TypeCobolChecker.cs]: Even better, target the token of the name of the function instead of the `DECLARE` keyword. you can retrieve it with `header.FunctionName.NameLiteral.Token`.

File: TypeCobol.Test/Parser/Programs/TypeCobol/ProcedureCall.rdzPGM.txt [hunk 6]
```diff
@@ -7,9 +7,9 @@ Line 37[21,26] <27, Error, Syntax> - Syntax error : Parameter with name 'format'
 Line 38[21,24] <27, Error, Syntax> - Syntax error : Parameter with name 'var1' declared multiple times OffendingSymbol=[21,24:var1]<UserDefinedWord>
 Line 39[21,24] <27, Error, Syntax> - Syntax error : Parameter with name 'var1' declared multiple times OffendingSymbol=[21,24:var1]<UserDefinedWord>
 Line 41[21,26] <27, Error, Syntax> - Syntax error : Parameter with name 'format' declared multiple times OffendingSymbol=[21,26:format]<UserDefinedWord>
-Line 111[8,55] <27, Error, Syntax> - Syntax error : A function "proc2" with the same profile already exists in namespace "ProcedureCall".
-Line 115[8,55] <27, Error, Syntax> - Syntax error : A function "proc2" with the same profile already exists in namespace "ProcedureCall".
-Line 119[8,70] <27, Error, Syntax> - Syntax error : A function "proc2" with the same profile already exists in namespace "ProcedureCall".
+Line 111[8,14] <27, Error, Syntax> - Syntax error : A function "proc2" with the same profile already exists in namespace "ProcedureCall". OffendingSymbol=[8,14:declare]<DECLARE>
```
Comment (fm-117) [inline TypeCobol.Test/Parser/Programs/TypeCobol/ProcedureCall.rdzPGM.txt]: Should not it be a Semantic error here ?

Comment (LANOYDO) [reply TypeCobol.Test/Parser/Programs/TypeCobol/ProcedureCall.rdzPGM.txt]: done.
https://github.com/TypeCobolTeam/TypeCobol/pull/1669/commits/4b5e1d0af58323248efbd75657d4b554eb7db081

File: TypeCobol/Compiler/Diagnostics/TypeCobolChecker.cs (hunk 5 above)
Comment (LANOYDO) [reply TypeCobol/Compiler/Diagnostics/TypeCobolChecker.cs]: done.
https://github.com/TypeCobolTeam/TypeCobol/pull/1669/commits/4b5e1d0af58323248efbd75657d4b554eb7db081

File: TypeCobol.LanguageServer.Test/LSRTests/ReplacingSyntaxOff/input/replacingSyntaxOff.tlsp (hunk 4 above)
Comment (LANOYDO) [reply TypeCobol.LanguageServer.Test/LSRTests/ReplacingSyntaxOff/input/replacingSyntaxOff.tlsp]: done.
https://github.com/TypeCobolTeam/TypeCobol/pull/1669/commits/4b5e1d0af58323248efbd75657d4b554eb7db081

File: TypeCobol.LanguageServer.Test/LSRTests/MultiLineCommentIncScan/input/MultiLineCommentIncScan.tlsp (hunk 3 above)
Comment (LANOYDO) [reply TypeCobol.LanguageServer.Test/LSRTests/MultiLineCommentIncScan/input/MultiLineCommentIncScan.tlsp]: fixed.
https://github.com/TypeCobolTeam/TypeCobol/pull/1669/commits/4b5e1d0af58323248efbd75657d4b554eb7db081

File: TypeCobol.LanguageServer.Test/LSRTests/MultiCommentSyntaxColoring/input/MultiCommentSyntaxColoring.tlsp (hunk 2 above)
Comment (LANOYDO) [reply TypeCobol.LanguageServer.Test/LSRTests/MultiCommentSyntaxColoring/input/MultiCommentSyntaxColoring.tlsp]: done.
https://github.com/TypeCobolTeam/TypeCobol/pull/1669/commits/4b5e1d0af58323248efbd75657d4b554eb7db081

File: TypeCobol/Compiler/Diagnostics/TypeCobolChecker.cs (hunk 1 above)
Comment (LANOYDO) [reply TypeCobol/Compiler/Diagnostics/TypeCobolChecker.cs]: Remove in grammar possibility of condition in procedure parameter.
https://github.com/TypeCobolTeam/TypeCobol/pull/1669/commits/77d03737d977c6907c9a5d2deea5d19ffd2cd74f

Comment (smedilol) [reply TypeCobol/Compiler/Diagnostics/TypeCobolChecker.cs]: Can you also remove the related C# code ?

Comment (LANOYDO) [reply TypeCobol/Compiler/Diagnostics/TypeCobolChecker.cs]: done
https://github.com/TypeCobolTeam/TypeCobol/pull/1669/commits/39f4a7a95c122ab7a4af7c18a622d1a8679e1d5b

File: TypeCobol.Test/Parser/Programs/Cobol2002/LevelChecker.rdz.tcbl [hunk 7]
```diff
@@ -40,10 +40,4 @@
 
        PROCEDURE DIVISION.
 
-      *KO Level 88 under type bool in parameter
-       DECLARE PROCEDURE MyProc
-         INPUT exists TYPE BOOL
-                88 exists-Ok value 'O'.
-       END-DECLARE.
```
Comment (fm-117) [inline TypeCobol.Test/Parser/Programs/Cobol2002/LevelChecker.rdz.tcbl]: I think we should keep the test and simply update the error message. It is still an invalid syntax, only the way to check it has changed.

File: TypeCobol.Test/Parser/Programs/TypeCobol/FunctionsDeclaration.rdz.tcbl [hunk 8]
```diff
@@ -111,9 +111,6 @@
        
        DECLARE FUNCTION FunConditions PRIVATE
              INPUT  gender PIC X(01)
-                 88  valid-gender VALUE 'F' 'M'
-                 88  female VALUE 'F'
-                 88  male   VALUE 'M'
```
Comment (fm-117) [inline TypeCobol.Test/Parser/Programs/TypeCobol/FunctionsDeclaration.rdz.tcbl]: We should keep the test here too

File: TypeCobol.Test/Parser/Programs/TypeCobol/FunctionsDeclaration.rdz.tcbl [hunk 9]
```diff
@@ -122,9 +119,6 @@
       * ERROR only level-88 parameter items shall have an explicit level number
        DECLARE FUNCTION FunConditions PRIVATE
              INPUT 88 valid-gender VALUE 'F' 'M'
-                      gender PIC X(01)
-                   88  female VALUE 'F'
-                   01  male   VALUE 'M'
```
Comment (fm-117) [inline TypeCobol.Test/Parser/Programs/TypeCobol/FunctionsDeclaration.rdz.tcbl]: Same here, keep this invalid syntax test. By removing it, you'll have an unintended conflict with the previous FunConditions declaration. The idea is to update the error messages but not necessarily change the test data.

File: TypeCobol/AntlrGrammar/TypeCobolCodeElements.g4 [hunk 10]
```diff
@@ -104,7 +104,8 @@ functionReturningPhrase: RETURNING parameterDescription;
 // parameterDescription is a rule created from dataDescriptionEntry and enforcing rules:
 // - TCRFUN_PARAMETER_DESCRIPTION
 // - TCRFUN_LEVEL_88_PARAMETERS
-parameterDescription: (functionDataParameter | functionConditionParameter) PeriodSeparator?;
+// parameterDescription: (functionDataParameter | functionConditionParameter) PeriodSeparator?;
+parameterDescription: functionDataParameter PeriodSeparator?;
```
Comment (fm-117) [inline TypeCobol/AntlrGrammar/TypeCobolCodeElements.g4]: No need to keep the previous rule definitions in comments, you can delete them. Also please check that there is no more references to `TCRFUN_LEVEL_88_PARAMETERS` in code.

File: TypeCobol/AntlrGrammar/TypeCobolCodeElements.g4 [hunk 11]
```diff
@@ -138,8 +139,8 @@ tcfuncParameterUsageClause:
 					//FUNCTION_POINTER and PROCEDURE_POINTER are not implemented yet
 				    ;
 
-functionConditionParameter:
-	levelNumber=IntegerLiteral conditionNameDefinition valueClauseForCondition;
+//functionConditionParameter:
```
Comment (fm-117) [inline TypeCobol/AntlrGrammar/TypeCobolCodeElements.g4]: same as above

File: TypeCobol.Test/Parser/Programs/Cobol2002/LevelChecker.rdz.tcbl (hunk 7 above)
Comment (LANOYDO) [reply TypeCobol.Test/Parser/Programs/Cobol2002/LevelChecker.rdz.tcbl]: done
https://github.com/TypeCobolTeam/TypeCobol/pull/1669/commits/3c0f9d082a18ff1b641cb676ce9152a389f4f984

File: TypeCobol.Test/Parser/Programs/TypeCobol/FunctionsDeclaration.rdz.tcbl (hunk 8 above)
Comment (LANOYDO) [reply TypeCobol.Test/Parser/Programs/TypeCobol/FunctionsDeclaration.rdz.tcbl]: done
https://github.com/TypeCobolTeam/TypeCobol/pull/1669/commits/3c0f9d082a18ff1b641cb676ce9152a389f4f984

File: TypeCobol.Test/Parser/Programs/TypeCobol/FunctionsDeclaration.rdz.tcbl (hunk 9 above)
Comment (LANOYDO) [reply TypeCobol.Test/Parser/Programs/TypeCobol/FunctionsDeclaration.rdz.tcbl]: done
https://github.com/TypeCobolTeam/TypeCobol/pull/1669/commits/3c0f9d082a18ff1b641cb676ce9152a389f4f984

File: TypeCobol/AntlrGrammar/TypeCobolCodeElements.g4 (hunk 10 above)
Comment (LANOYDO) [reply TypeCobol/AntlrGrammar/TypeCobolCodeElements.g4]: done
https://github.com/TypeCobolTeam/TypeCobol/pull/1669/commits/3c0f9d082a18ff1b641cb676ce9152a389f4f984

File: TypeCobol/AntlrGrammar/TypeCobolCodeElements.g4 (hunk 11 above)
Comment (LANOYDO) [reply TypeCobol/AntlrGrammar/TypeCobolCodeElements.g4]: done
https://github.com/TypeCobolTeam/TypeCobol/pull/1669/commits/3c0f9d082a18ff1b641cb676ce9152a389f4f984

File: TypeCobol.LanguageServer.Test/LSRTests/ProcedureCompletionInputInoutOutput-NoContext/input/ProcedureCompletionInputInoutOutput-NoContext.tlsp [hunk 12]
```diff
@@ -131,7 +131,7 @@
     },
     {
       "category": 2,
-      "message": "{\"jsonrpc\":\"2.0\",\"id\":\"2\",\"result\":[{\"label\":\"BirthDate (DATE) (Person::BirthDate)\",\"kind\":6,\"insertText\":\"Person::BirthDate IN-OUT \"},{\"label\":\"MyDate (DATE) (MyDate)\",\"kind\":6,\"insertText\":\"MyDate IN-OUT \"}]}"
+      "message": "{\"jsonrpc\":\"2.0\",\"id\":\"2\",\"result\":[{\"label\":\"BirthDate (DATE) (Person::BirthDate)\",\"kind\":6,\"insertText\":\"Person::BirthDate\"},{\"label\":\"MyDate (DATE) (MyDate)\",\"kind\":6,\"insertText\":\"MyDate\"}]}"
```
Comment (fm-117) [inline TypeCobol.LanguageServer.Test/LSRTests/ProcedureCompletionInputInoutOutput-NoContext/input/ProcedureCompletionInputInoutOutput-NoContext.tlsp]: In this file, the completion results have changed. This is normal to have changes in error messages (textDocument/publishDiagnostics) but a bit strange to have changes in completion results. The grammar changed so I guess that the produced CodeElement/Nodes are now different and this affects the completion process. We should check whetehr the results are ok or not.

File: TypeCobol.LanguageServer.Test/TypeCobol.LanguageServer.Test.csproj [hunk 13]
```diff
@@ -69,398 +69,7 @@
...
-      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
-    </None>
-    <None Include="LSRTests\VariableQualifiedNameCompletion2\output_expected\VaraibleQualifiedNameCompletion2.rlsp">
-      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
-    </None>
-    <None Include="LSRTests\VariableQualifiedNameCompletion\input\VariableQualifiedSeparatorCompletion.tlsp">
-      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
-    </None>
-    <None Include="LSRTests\VariableQualifiedNameCompletion\output_expected\VariableQualifiedSeparatorCompletion.rlsp">
+	<None Include="LSRTests\**">
```
[the remark this answers is not in the document]
Comment (LANOYDO) [reply TypeCobol.LanguageServer.Test/TypeCobol.LanguageServer.Test.csproj]: done
https://github.com/TypeCobolTeam/TypeCobol/pull/1669/commits/aa9a53a6fc86e6e87964a13ad3dc79e519e409fb

File: TypeCobol.LanguageServer.Test/LSRTests/ProcedureCompletionInputInoutOutput-NoContext/input/ProcedureCompletionInputInoutOutput-NoContext.tlsp (hunk 12 above)
Comment (LANOYDO) [reply TypeCobol.LanguageServer.Test/LSRTests/ProcedureCompletionInputInoutOutput-NoContext/input/ProcedureCompletionInputInoutOutput-NoContext.tlsp]: Check on branch v1.4.1 displays the same diffene

**Comment by smedilol** [comment]
Cancel this PR.
Replaced by #1691
